src/object/obj_class.c

```diff
@@ -779,3 +773,99 @@ obj_encode_full_stripe(daos_obj_id_t oid, d_sg_list_t *sgl, uint32_t *sg_idx,
 		D_FREE(ldata[i]);
 	return rc;
 }
+
+static void
+oc_sop_swap(void *array, int a, int b)
+{
+	struct daos_obj_class **ocs = (struct daos_obj_class **)array;
+	struct daos_obj_class  *tmp;
+
+	tmp = ocs[a];
+	ocs[a] = ocs[b];
+	ocs[b] = tmp;
+}
+
+static int
+oc_sop_cmp(void *array, int a, int b)
+{
+	struct daos_obj_class **ocs = (struct daos_obj_class **)array;
+
+	if (ocs[a]->oc_id > ocs[b]->oc_id)
+		return 1;
+	if (ocs[a]->oc_id < ocs[b]->oc_id)
+		return -1;
+	return 0;
+}
+
+static int
+oc_sop_cmp_key(void *array, int i, uint64_t key)
+{
+	struct daos_obj_class **ocs = (struct daos_obj_class **)array;
+	unsigned int		id  = (unsigned int)key;
+
+	if (ocs[i]->oc_id > id)
+		return 1;
+	if (ocs[i]->oc_id < id)
+		return -1;
+	return 0;
+}
+
+static daos_sort_ops_t	oc_sort_ops = {
+	.so_swap	= oc_sop_swap,
+	.so_cmp		= oc_sop_cmp,
+	.so_cmp_key	= oc_sop_cmp_key,
+};
+
+/* NB: ignore the last one which is UNKNOWN */
+#define OC_NR	ARRAY_SIZE(daos_obj_classes)
+
+static struct daos_obj_class *
+obj_class_id2cl(int oc_id)
+{
+	int	idx;
+
+	if (oc_id == OC_UNKNOWN)
+		return NULL;
+
+	idx = daos_array_find(obj_classes, OC_NR, oc_id, &oc_sort_ops);
+	if (idx < 0)
+		return NULL;
+
+	return obj_classes[idx];
+}
+
+int
+obj_class_init(void)
+{
+	int	i;
+	int	rc;
+
+	if (obj_classes)
+		return 0;
+
+	D_ALLOC_ARRAY(obj_classes, OC_NR);
```

Comment:
It caused NLT error. This looks not right, obj_classes should be "struct daos_obj_class *".

Reply:
my fault, seems your patch uses pointer array, the NLT error is about "obj_class" is not freed, probably it because module fini function isn't called anywhere?